addons/l10n_fr_hr_work_entry_holidays/models/hr_contract.py

```diff
@@ -0,0 +1,72 @@
+        # l10n_fr_date_to is False when no adjustment had to be done
+        all_leaves = self.env['hr.leave'].search([
+            ('employee_id', 'in', fr_contracts.employee_id.ids),
+            ('state', '=', 'validate'),
+            ('date_from', '<=', datetime_to_string(end_dt)),
+            ('l10n_fr_date_to', '>=', datetime_to_string(start_dt)),
+        ]) if fr_contracts else []
+        leaves_per_employee = defaultdict(lambda: self.env['hr.leave'])
+        for leave in all_leaves:
+            leaves_per_employee[leave.employee_id] |= leave
+        for contract in fr_contracts:
+            if contract.company_id.country_id.code != 'FR':
+                continue
+            employee = contract.employee_id
+            employee_calendar = contract.resource_calendar_id
+            company = contract.company_id
+            company_calendar = company.resource_calendar_id
+            resource = employee.resource_id
+            tz = pytz.timezone(employee_calendar.tz)
+
+            for leave in leaves_per_employee[employee]:
+                leave_start_dt = max(start_dt, leave.date_from.astimezone(tz))
+                leave_end_dt = min(end_dt, leave.date_to.astimezone(tz))
+                leave_end_dt_fr = min(end_dt, leave.l10n_fr_date_to.astimezone(tz))
+
+                # Compute the attendances for the company calendar and the employee calendar
+                # and then compute and keep the difference between those two
+                employee_attendances = employee_calendar._attendance_intervals_batch(
+                    leave_start_dt, leave_end_dt, resources=resource, tz=tz,
+                )[resource.id]
+                company_attendances = company_calendar._attendance_intervals_batch(
+                    leave_start_dt, leave_end_dt_fr, resources=resource, tz=tz,
+                )[resource.id]
+                # Dates on which work entries should already be generated
+                employee_dates = set()
+                for interval in employee_attendances:
+                    employee_dates.add(interval[0].date())
+                    employee_dates.add(interval[1].date())
+                leave_work_entry_type = leave.holiday_status_id.work_entry_type_id
+                result += [{
+                    'name': '%s%s' % (leave_work_entry_type.name + ': ' if leave_work_entry_type else "", employee.name),
+                    'date_start': interval[0].astimezone(pytz.utc).replace(tzinfo=None),
+                    'date_stop': interval[1].astimezone(pytz.utc).replace(tzinfo=None),
+                    'work_entry_type_id': leave_work_entry_type.id,
+                    'employee_id': employee.id,
+                    'company_id': contract.company_id.id,
+                    'state': 'draft',
+                    'contract_id': contract.id,
+                    'leave_id': leave.id,
+                } for interval in company_attendances if interval[0].date() not in employee_dates]
```

Comment:
Looking back at this I think we can avoid the first call to `_attendance_intervals_batch` by constructing employee_dates using the values in the result
```suggestion
                company_attendances = company_calendar._attendance_intervals_batch(
                    leave_start_dt, leave_end_dt_fr, resources=resource, tz=tz,
                )[resource.id]
                # Dates on which work entries should already be generated
                employee_dates = set()
                for vals in result:
                    employee_dates.add(vals['date_start'].date())
                    employee_dates.add(vals['date_stop'].date())
                leave_work_entry_type = leave.holiday_status_id.work_entry_type_id
                result += [{
                    'name': '%s%s' % (leave_work_entry_type.name + ': ' if leave_work_entry_type else "", employee.name),
                    'date_start': interval[0].astimezone(pytz.utc).replace(tzinfo=None),
                    'date_stop': interval[1].astimezone(pytz.utc).replace(tzinfo=None),
                    'work_entry_type_id': leave_work_entry_type.id,
                    'employee_id': employee.id,
                    'company_id': contract.company_id.id,
                    'state': 'draft',
                    'contract_id': contract.id,
                    'leave_id': leave.id,
                } for interval in company_attendances if interval[0].date() not in employee_dates]
```
Technically we can also avoid calling `_attendance_intervals_batch` for each leave as well using that same logic but I am not sure which version would be less costly, it is also not that likely to have a lot of leaves.